### ISSUE TITLE
systemd-coredump: allow reverting to traditional, local directory coredump behaviour selectively and locally for specific services

### ISSUE THREAD

**Issue by aarnez** [systemd/systemd#7166]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

> 231

### Used distribution

> Fedora 25

### Request

When systemd-coredump is active, it would sometimes be preferable to
exempt certain user programs from the journal processing.  For instance,
consider a developer on a multi-user system whose program frequently
crashes during a debug session.  The system administrator would probably
not be interested in these crashes.  And most of the resulting core files
can likely be deleted soon after their creation.  In that case it is more
appropriate to dump the core file into the user's directory and suppress
the journal entry, such that:

- no system journal flooding occurs;
- core dump size and processing time can be fully accounted to the user;
- no extra step (such as coredumpctl) is needed for dump retrieval;
- core file dumping and processing is more easily scriptable.

This would particularly benefit use cases such as the GDB test suite,
which deliberately generates multiple core files and performs various
tests with GDB on them.  Currently the GDB test suite is not prepared to
execute core file tests when systemd-coredump is active.  Invoking
coredumpctl there has been discussed and considered troublesome:
https://sourceware.org/ml/gdb-patches/2017-10/msg00537.html

An open question is how to switch this "classical" core dump behavior on
or off.  One option may be a user-controllable per-process setting --
something like "ulimit -c unlimited" or an environment variable.  Another
might be a system-level configuration option that enables "classical"
handling based on characteristics of the crashed program, e.g. owner,
path, which namespaces it is associated with, etc.

**Comment by aarnez** [comment]
Any thoughts on this?

**Comment by poettering** [comment]
i don't think maintaining a blacklist like this in journald.conf is acceptable.

That said, it would be OK to have a per-process opt-out of this logic, but I wouldn't know a safe way to pass this information to the coredump logic. After all, it's not really safe to make the kernel dump coredumps to arbitrary directories, hence it needs to be a privileged option to turn this logic on, but afaics there's simply no fitting per-process functionality in the kernel that would allow us to attach things to a process like that.

**Comment by boucman** [comment]
hmm,couldn't that be wired to logind and session management ? if a process that dumps core is part of a user session, dump it in a user-related subdirectory instead of the more global one...

logind could create some sort of config dropins for systemd-coredump in /run...

IIRC it's possible to know if a pid is part of a user session using cgroups, but I'm a bit out of my comfort area here...

**Comment by boucman** [comment]
pushing the idea a bit more...

how about having a per-user systemd-coredump socket in /run... if the socket exist and the process is part of a session, use that socket, thus starting a user-instance of systemd-coredump. if it doesn't, then use the system socket, thus reverting to the default behaviour...

**Comment by poettering** [comment]
> if a process that dumps core is part of a user session, dump it in a user-related subdirectory instead of the more global one...

That doesn't sound desirable to me. Dumping core into the working directory is a developer thing, and definitely nothing that should apply to normal user processes. If gnome-shell dies, then yes, systemd-coredump should deal with that, absolutely.

**Comment by boucman** [comment]
yeah, I like my second idea better, having a way for a user to specify that he wants to deal with his coredumps himself, and if that is not set, fallback to the system-wide configuration for coredumps.

I'm not sure how feasible it is, but i'll try to have a look at some point...

**Comment by aarnez** [comment]
On Wed, Dec 13 2017, Boucman wrote:

> [...] having a way for a user to specify that he wants to deal with
> his coredumps himself [...]

Maybe systemd-coredump could evaluate certain environment variables of
the dumped process, e.g.:

    SYSTEMD_CORE_PATTERN=~/my-corefiles/core.%p

But I think we could also establish a set of characteristics of a dumped
process, such as "stdin is a tty", "has a shell as ancestor", or "soft
core file size limit is unlimited".  And then use those for defining a
rule like

    TraditionalRule=(stdin_is_tty | has_shell_ancestor) & core_size_unlimited

in `coredump.conf`.  This potentially relieves the individual users from
bothering with additional config options in their development/debug
sessions.

Just a thought.

**Comment by poettering** [comment]
well, i am pretty sure this needs to be something only a privileged process can enable really, otherwise it'll be a security problem, given that it will be a more privileged process that writes to arbitrary user directories...

**Comment by aarnez** [comment]
Hm, couldn't we setuid() to the owner of the crashed process before writing the dump?

**Comment by jfilak** [comment]
Dumping a core to a user specified directory is not as easy as it might seem to be. We have had hard times with this functionality in ABRT. You must not forget about SELinux and User namespaces and ton of other little things. Here is the function that opens a user core dump file:
https://github.com/abrt/abrt/blob/master/src/hooks/abrt-hook-ccpp.c#L149

**Comment by aarnez** [comment]
Thanks for the info.  Note that abrt-ccpp has obviously been disabled by
default on Fedora distros since Fedora 26:
https://abrt.github.io/abrt/systemd/core_pattern/2017/03/07/fedora-26-change/

So, could (some of) this logic be re-used for systemd-coredump then?  Or
is there a better way of re-establishing this functionality?

**Comment by jfilak** [comment]
Well, I do not believe that systemd-coredump will ever gain this functionality. Systemd maintainers keep the position that you should either get coredump files from journald (like abrt-dump-journal-core) or change the core_pattern (Fedora will always ship abrt-ccpp-hook for those who want the functionality it provides).

Frankly, I understand them - systemd-coredump is here to save core files in journal (with exceptions, of course) and if you need something else you should use a different tool.
